File: leetcode/easy/111=minimumDepthOfBinaryTree.java
// first pass solution
class Solution {
    int min = Integer.MAX_VALUE;
    
    public int minDepth(TreeNode root) {
        if (root == null) return 0;
        
        DFS(root, 1);
        
        return min;
    }
    
    public void DFS(TreeNode node, int level) {
        if (node.left == null && node.right == null) min = Math.min(min, level);
        if (node.left != null) DFS(node.left, level + 1);
        if (node.right != null) DFS(node.right, level + 1);
    }
}

// second pass solution
class Solution {
    public int minDepth(TreeNode root) {
        if (root == null) return 0;
        
        return DFS(root, 1);
    }
    
    public int DFS(TreeNode node, int level) {
        if (node == null) return Integer.MAX_VALUE;
        
        if (node.left == null && node.right == null) return level;
        
        int left = DFS(node.left, level + 1);
        int right = DFS(node.right, level + 1);
        
        return Math.min(left, right);
    }
}

// third pass solution
class Solution {
    public int minDepth(TreeNode root) {
        if (root == null) return 0;
        
        List<TreeNode> queue = new ArrayList<>();
        int output = Integer.MAX_VALUE;
        int currLevel = 1;
        
        queue.add(root);
        
        while (queue.size() > 0) {
            int len = queue.size();

            for (int i = 0; i < len; i++) {
                TreeNode currNode = queue.remove(0);

                if (currNode.left == null && currNode.right == null) return currLevel;
                if (currNode.left != null) queue.add(currNode.left);
                if (currNode.right != null) queue.add(currNode.right);
            }
            currLevel++;
        }
        return currLevel;
    }
}

File: leetcode/easy/111=minimumDepthOfBinaryTree.js
// Given a binary tree, find its minimum depth.

// The minimum depth is the number of nodes along the shortest path from the root node down to the nearest leaf node.

// Note: A leaf is a node with no children.

// Example 1:

// Input: root = [3,9,20,null,null,15,7]
// Output: 2
// Example 2:

// Input: root = [2,null,3,null,4,null,5,null,6]
// Output: 5

// first pass iterative solution using BST/queue
var minDepth = function(root) {
    if (!root) return 0;
    
    let queue = [[root, 1]];
    
    while (queue.length > 0) {
        let [node, depth] = queue.shift();
        
        if (!node.left && !node.right) {
            return depth;
        }
        
        if (node.left) queue.push([node.left, depth + 1]);
        if (node.right) queue.push([node.right, depth + 1]);
    }
};

// second pass recursive solution
var minDepth = function(root) {
    if (!root) return 0;
    else if (root && !root.left && !root.right) return 1;
    else if (root.left && !root.right) return minDepth(root.left) + 1;
    else if (root.right && !root.left) return minDepth(root.right) + 1;
    else return Math.min(minDepth(root.left) + 1, minDepth(root.right) + 1);
};
